rclpy/src/rclpy/qos.cpp

```diff
@@ -0,0 +1,135 @@
+// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Include pybind11 before rclpy_common/handle.h includes Python.h
+#include <pybind11/pybind11.h>
+
+#include <rcl/rcl.h>
+#include <rcl/time.h>
+#include <rmw/types.h>
+
+#include <cstring>
+#include <memory>
+#include <string>
+
+#include "rclpy_common/common.h"
+#include "rclpy_common/handle.h"
+
+#include "qos.hpp"
+
+namespace rclpy
+{
+
+namespace
+{
+
+// Fill a given rmw_time_t from an rclpy.duration.Duration instance.
+void
+_convert_py_duration_to_rmw_time(py::capsule pyduration, rmw_time_t * out_time)
+{
+  assert(out_time != nullptr);
+  auto duration = static_cast<rcl_duration_t *>(pyduration);
```

Comment:
I'd recommend doing a `std::strcmp` on the capsule name to make sure this is a duration capsule.

Example: https://github.com/ros2/rclpy/blob/8a712bda9ebf1e7c5b3c94180584e6d687d9725d/rclpy/src/rclpy/client.cpp#L76-L79

Reply:
Ah good catch! See https://github.com/ros2/rclpy/pull/736/commits/e0d364465df44d78bd4fade06e65bb1be177467f.